### ISSUE TITLE
Is this really the correct implementation of the paper?

### ISSUE THREAD

**Issue by ZorkroZ** [felixlaumon/deform-conv#4]
It seems to me that the type of deformation learned is different from the one described in the publication.

Of what I understand of your implementation you learn 1 vector offset for each of your filters in the ConvOffset2D layer, so you just learn nb_filters translations and apply them 1 to 1 to the channels of the input. This is a big limitation to the potential deformations learned by the Deformable conv net of the paper (like rotations).

If I did not really grasp all the ins and outs this implementation, please enlighten me.

**Comment by HC-2016** [comment]
I agree that the implementation is different from the original paper. 
It seems that ConvOffset2D learns to rearrange the "pixels" in each channel severally.

But I'm confused with the following code:
`   coords_offset_lt = coords - tf.cast(coords_lt, 'float32')
    vals_t = vals_lt + (vals_rt - vals_lt) * coords_offset_lt[..., 0]
    vals_b = vals_lb + (vals_rb - vals_lb) * coords_offset_lt[..., 0]
    mapped_vals = vals_t + (vals_b - vals_t) * coords_offset_lt[..., 1]`

I think the code should be corrected as 
`   coords_offset_lt = coords - tf.cast(coords_lt, 'float32')
    vals_t = vals_lt + (vals_rt - vals_lt) * coords_offset_lt[..., 1]
    vals_b = vals_lb + (vals_rb - vals_lb) * coords_offset_lt[..., 1]
    mapped_vals = vals_t + (vals_b - vals_t) * coords_offset_lt[..., 0]`

**Comment by rouniuyizu** [comment]
I have similar questions on this. According to the paper (P2, starting section 2): "While the feature maps and convolution in CNNs are 3D, both deformable convolution and RoI pooling operate on the 2D spatial domain and **remain the same across the channel dimension**." My take of it is all the offsets shall be the same across different channels. For example in RGB world, if you have (-1, -1) offset selected on R channel, it shall also apply to G and B, while the code here learns different offset per channel which somehow worries me...

**Comment by HC-2016** [comment]
Obviously this code is a simplified version of original paper. 
The implementation for original paper  [https://github.com/msracver/Deformable-ConvNets/tree/master/rfcn/operator_cxx](url) involves the modification of im2col.

**Comment by vyokky** [comment]
Agree with above. What I understand is that the number of offset channel should be 2*f_w*f_h, where f_w and f_h are the width and height of the filter. For example, if the filter is 3*3, then the offset channel should be 18. Please enlighten me if I am wrong.

**Comment by skbhat** [comment]
I think the comparison in scaled_mnist.py is unfair.  First model cnn.h5 is learned on regular training images.  Then deform_cnn.h5 is initiated with the weights of cnn.h5 and trained using scaled dataset.  Then the performance of cnn.h5 and deform_cnn.h5 are compared on the scaled dataset.  But cnn.h5 was never trained on scaled dataset.

When I train cnn5.h5 on scaled dataset I get 98.8% accuracy.  Please add this code and try yourself

inputs, outputs = get_cnn()
model = Model(inputs=inputs, outputs=outputs)
model.load_weights('models/cnn.h5', by_name=True)
model.summary()
optim = Adam(5e-4)
# optim = SGD(1e-4, momentum=0.99, nesterov=True)
loss = categorical_crossentropy
model.compile(optim, loss, metrics=['accuracy'])

model.fit_generator(
    train_scaled_gen, steps_per_epoch=steps_per_epoch,
    epochs=20, verbose=1,
    validation_data=test_scaled_gen, validation_steps=validation_steps
)
model.save_weights('models/cnn_regular_trained_on_scaled.h5')


model.load_weights('models/cnn_regular_trained_on_scaled.h5')

val_loss, val_acc = model.evaluate_generator(
    test_gen, steps=validation_steps
)
print('Test accuracy of regular convolution with regular images', val_acc)

val_loss, val_acc = model.evaluate_generator(
    test_scaled_gen, steps=validation_steps
)
print('Test accuracy of regular convolution with scaled images', val_acc)

**Comment by XuanheLiu** [comment]
@vyokky I agree with you. I think the code only takes an offset value in each location, but it should correspond to the 3*3=9 offset value in each location, which should correspond to the 9 level offset value.Can we talk about it? My WeChat1538548490

**Comment by ChunhuanLin** [comment]
I think there are really some differences from the original paper, hence I have my own PyTorch implementation [here](https://github.com/ChunhuanLin/deform_conv_pytorch).

**Comment by isaacgerg** [comment]
I agree, this is not the implementation in the paper.

**Comment by tianzhuotao** [comment]
@HC-2016 
Have you tried the modification you mentioned on this issue ? Does the result get better ?  I have not read the original paper in details yet.
Thanks.

**Comment by CLegendBuptSun** [comment]
Obviously，the implementation is different from the original paper. I agree with @vyokky and @HC-2016. The ConvOffset2D op is just learning to rearrange the "pixels" in each channel severally.
1. We could treat it as a specific-specific mode of deformable convs. 
2.Channel-wise transformation is different from paper. The deformable op is just applied to kernels through all channels.

**Comment by CLegendBuptSun** [comment]
So I don't know why there are so many stars. 
This could mislead the beginner.
Please enlighten me if I am wrong.
